<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>nxrepair &mdash; misassembly_detection 1.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="misassembly_detection 1.0 documentation" href="../index.html" />
    <link rel="up" title="Module code" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">misassembly_detection 1.0 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for nxrepair</h1><div class="highlight"><pre>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">from</span> <span class="nn">intervalNode</span> <span class="kn">import</span> <span class="n">IntervalNode</span>


<span class="k">def</span> <span class="nf">meansd</span><span class="p">(</span><span class="n">frq</span><span class="p">):</span>

    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Function to calculate mean and standard deviation from a dictionary of frequencies.</span>

<span class="sd">    Return a tuple of (mean, std).</span>

<span class="sd">    Arguments:</span>
<span class="sd">    frq: a dictionary of frequencies: key = insert size, value = frequency</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">keys</span> <span class="o">=</span> <span class="n">frq</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="n">keys</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="n">w</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">keys</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">float</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">k</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">keys</span><span class="p">):</span>
        <span class="n">w</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">frq</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>

    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">keys</span><span class="p">)</span>
    <span class="n">xbar</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">average</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">weights</span><span class="o">=</span><span class="n">w</span><span class="p">)</span>
    <span class="n">xsd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">average</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">power</span><span class="p">((</span><span class="n">x</span> <span class="o">-</span> <span class="n">xbar</span><span class="p">),</span><span class="mi">2</span><span class="p">)),</span><span class="n">weights</span><span class="o">=</span><span class="n">w</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">xbar</span><span class="p">,</span><span class="n">xsd</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">MAD</span><span class="p">(</span><span class="n">frq</span><span class="p">):</span>

    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Function to calculate median and median absolute deviation (MAD) from a dictionary of frequencies.</span>

<span class="sd">    Return a tuple of (mean, MAD).</span>

<span class="sd">    Arguments:</span>
<span class="sd">    frq: a dictionary of frequencies: key = insert size, value = frequency</span>
<span class="sd">     </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">all_lengths</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">frq</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="n">new_vals</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">*</span> <span class="nb">int</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
        <span class="n">all_lengths</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">new_vals</span><span class="p">)</span>
    <span class="n">all_lengths</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">all_lengths</span><span class="p">))</span>
    <span class="c">#all_lengths = np.log(all_lengths)</span>
    <span class="n">mid</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_lengths</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
    <span class="n">median</span> <span class="o">=</span> <span class="n">all_lengths</span><span class="p">[</span><span class="n">mid</span><span class="p">]</span>
    <span class="n">residuals</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">all_lengths</span> <span class="o">-</span> <span class="n">median</span><span class="p">))</span>
    <span class="n">MAD</span> <span class="o">=</span> <span class="n">residuals</span><span class="p">[</span><span class="n">mid</span><span class="p">]</span>
    <span class="k">print</span> <span class="n">MAD</span>
    <span class="k">return</span> <span class="n">median</span><span class="p">,</span> <span class="n">MAD</span>

<span class="k">def</span> <span class="nf">find_intersections</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">s</span><span class="p">,</span> <span class="n">e</span><span class="p">):</span>

    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Function to find inserts that bridge a region of a contig.</span>

<span class="sd">    Arguments:</span>
<span class="sd">    tree: interval tree of start/end positions of mate pair inserts</span>
<span class="sd">    s: start position of interval on contig </span>
<span class="sd">    e: end position of interval on contig</span>
<span class="sd">     </span>
<span class="sd">    Return a list of nodes from tree whose start and end positions span the interval given by s and e. </span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c"># find all reads that bridge a gap</span>
    <span class="n">intersections</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">e</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">intersections</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">intersections</span>

<span class="k">def</span> <span class="nf">get_insertlengths</span><span class="p">(</span><span class="n">reads</span><span class="p">):</span>

    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Function to calculate interval sizes of a set of mate pairs.</span>

<span class="sd">    Arguments:</span>
<span class="sd">    reads: a list of mate pairs as interal tree node objects</span>

<span class="sd">    Return two numpy arrays: an array of insert sizes (integers) and an array of strand alignments (boolean)  </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">distances</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">strands</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
        <span class="n">distances</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">read</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
        <span class="n">strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">other</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">distances</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">strands</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">probability_of_readlength</span><span class="p">(</span><span class="n">read_length</span><span class="p">,</span> <span class="n">mu</span><span class="p">,</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">pi1</span><span class="p">,</span> <span class="n">L</span><span class="p">):</span>

    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Function to calculate the probability that mate pair insert sizes are not anomalous.</span>
<span class="sd">    Return an array of probabilites.</span>

<span class="sd">    Arguments:</span>
<span class="sd">    read_length: a numpy array of insert sizes</span>
<span class="sd">    mu: mean insert size (float)</span>
<span class="sd">    sigma: insert size standard deviation (float)</span>
<span class="sd">    pi1: prior probability of being anomalous (float)</span>
<span class="sd">    L: length of contig to which the reads in read_length are aligned</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">p_0</span> <span class="o">=</span> <span class="n">pi1</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">L</span><span class="p">))</span>
    <span class="n">p_1</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">pi1</span><span class="p">)</span> <span class="o">*</span> <span class="n">stats</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">pdf</span><span class="p">(</span><span class="n">read_length</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="n">mu</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">sigma</span><span class="p">)</span>
    <span class="n">p_total</span> <span class="o">=</span> <span class="n">p_1</span> <span class="o">/</span> <span class="p">(</span><span class="n">p_0</span> <span class="o">+</span> <span class="n">p_1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">p_total</span>

<div class="viewcode-block" id="aligned_assembly"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly">[docs]</a><span class="k">class</span> <span class="nc">aligned_assembly</span><span class="p">:</span>

    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Class to hold a set of mate pair or paired end reads aligned to the scaffolded genome assembly</span>

<span class="sd">    &quot;&quot;&quot;</span>    

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bamfile</span><span class="p">,</span> <span class="n">fastafile</span><span class="p">,</span> <span class="n">min_size</span><span class="p">,</span> <span class="n">threshold</span><span class="p">,</span> <span class="n">step</span><span class="p">,</span> <span class="n">window</span><span class="p">,</span> <span class="n">minmapq</span><span class="p">,</span> <span class="n">maxinsert</span><span class="p">,</span> <span class="n">fraction</span><span class="p">,</span> <span class="n">prior</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Initialiser function for the aligned assembly class.</span>

<span class="sd">        Arguments:</span>
<span class="sd">        bamfile: a sorted bam file of reads aligned to the scaffolded assembly</span>
<span class="sd">        fastafile: the scaffolded assembly in fasta format</span>
<span class="sd">        min_size: the minimum size contig to consider for analysis (integer)</span>
<span class="sd">        threshold: threshold in Z score below which a misassembly is called (float)</span>
<span class="sd">        step: step size to walk contigs (integer)</span>
<span class="sd">        window: width of window around each position from which mate pair insert sizes are fetched (integer)</span>
<span class="sd">        minmapq: the minimum mapq value for which paired reads are evaluated (float)</span>
<span class="sd">        maxinsert: the maximum insert size for which genome population statistics are calculated </span>
<span class="sd">        fraction: minimum fraction of read pairs with correct orientation to call support for the assembly.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">minmapq</span> <span class="o">=</span> <span class="n">minmapq</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">maxinsert</span> <span class="o">=</span> <span class="n">maxinsert</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">step</span> <span class="o">=</span> <span class="n">step</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">window</span> <span class="o">=</span> <span class="n">window</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fraction</span> <span class="o">=</span> <span class="n">fraction</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">prior</span> <span class="o">=</span> <span class="n">prior</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sam</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">Samfile</span><span class="p">(</span><span class="n">bamfile</span><span class="p">,</span> <span class="s">&quot;rb&quot;</span> <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fasta</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">Fastafile</span><span class="p">(</span><span class="n">fastafile</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">min_size</span> <span class="o">=</span> <span class="n">min_size</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">all_reads</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sam</span><span class="o">.</span><span class="n">fetch</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">references</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sam</span><span class="o">.</span><span class="n">references</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sam</span><span class="o">.</span><span class="n">lengths</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">refdict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">read_stock</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">references</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">refdict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">=</span><span class="n">v</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">read_stock</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_reads</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span> 
        <span class="bp">self</span><span class="o">.</span><span class="n">sizes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_read_size_distribution</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">isize_mean</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">isize_sd</span> <span class="o">=</span> <span class="n">meansd</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sizes</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">isize_median</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">isize_MAD</span> <span class="o">=</span> <span class="n">MAD</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sizes</span><span class="p">)</span>
        <span class="k">print</span> <span class="bp">self</span><span class="o">.</span><span class="n">isize_sd</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">isize_MAD</span>

<div class="viewcode-block" id="aligned_assembly.get_read_size_distribution"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly.get_read_size_distribution">[docs]</a>    <span class="k">def</span> <span class="nf">get_read_size_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Function to calculate global insert size distribution across the whole assembly</span>
<span class="sd">        Return a frequency table of insert sizes as a  dictionary with key = insert size, value = frequency</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">frq</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">found</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">all_reads</span><span class="p">:</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">mapq</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">minmapq</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">rnext</span> <span class="o">==</span> <span class="n">read</span><span class="o">.</span><span class="n">tid</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">tlen</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">maxinsert</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">qname</span> <span class="ow">in</span> <span class="n">found</span> <span class="ow">and</span> <span class="n">found</span><span class="p">[</span><span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="n">read</span><span class="o">.</span><span class="n">tid</span><span class="p">:</span>
                    <span class="n">mate</span> <span class="o">=</span> <span class="n">found</span><span class="p">[</span><span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">]</span>
                    <span class="n">isize</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span> <span class="n">mate</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="n">mate</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">-</span><span class="n">read</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span><span class="n">read</span><span class="o">.</span><span class="n">pos</span><span class="o">+</span><span class="n">read</span><span class="o">.</span><span class="n">rlen</span><span class="o">-</span><span class="n">mate</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="p">))</span>
                    <span class="c">#isize = abs(read.tlen)</span>
                    <span class="n">frq</span><span class="p">[</span><span class="n">isize</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">found</span><span class="p">[</span><span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">tid</span><span class="p">,</span><span class="n">read</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span><span class="n">read</span><span class="o">.</span><span class="n">rlen</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">frq</span>
</div>
<div class="viewcode-block" id="aligned_assembly.get_reads"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly.get_reads">[docs]</a>    <span class="k">def</span> <span class="nf">get_reads</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Function to fetch reads aligned to a specific part of the assembled genome and return a list of aligned reads, where each list entry is a tuple:</span>
<span class="sd">        (read start position, read end position, read name, strand alignment) and strand alignment is a boolean indicating whether the two reads of a read pair align correctly to opposite strands.</span>
<span class="sd">        Reads are fetched that align to contig &quot;ref&quot; between positions &quot;start&quot; and &quot;end&quot;.</span>

<span class="sd">        Arguments:</span>
<span class="sd">        ref: the name of the contig from which aligned reads are to be fetched.</span>
<span class="sd">        start: the position on the contig from which to start fetching aligned reads</span>
<span class="sd">        end: the position on the contig from which to end fetching aligned reads</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c"># fetch all reads within a region, return those that are properly mapped</span>
        <span class="n">reads</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sam</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
        <span class="n">read_stock</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">found</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
            <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">rnext</span> <span class="o">==</span> <span class="n">read</span><span class="o">.</span><span class="n">tid</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">qname</span> <span class="ow">in</span> <span class="n">found</span> <span class="ow">and</span> <span class="n">found</span><span class="p">[</span><span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="n">read</span><span class="o">.</span><span class="n">tid</span><span class="p">:</span> <span class="c"># and (abs(read.tlen) &lt; 30000):</span>
                    <span class="n">mate</span> <span class="o">=</span> <span class="n">found</span><span class="p">[</span><span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">mate</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">read</span><span class="o">.</span><span class="n">pos</span><span class="p">:</span>
                        <span class="n">start_pos</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">read</span><span class="o">.</span><span class="n">rlen</span>
                        <span class="n">end_pos</span> <span class="o">=</span> <span class="n">mate</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">start_pos</span> <span class="o">=</span> <span class="n">mate</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">mate</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                        <span class="n">end_pos</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">pos</span>
                    <span class="n">correct_strands</span> <span class="o">=</span> <span class="p">((</span><span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">mate_is_reverse</span><span class="p">))</span> <span class="ow">or</span> <span class="p">((</span><span class="n">read</span><span class="o">.</span><span class="n">mate_is_reverse</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">))</span>
                    <span class="n">read_stock</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">start_pos</span><span class="p">,</span> <span class="n">end_pos</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">,</span> <span class="n">correct_strands</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">found</span><span class="p">[</span><span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">tid</span><span class="p">,</span><span class="n">read</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span><span class="n">read</span><span class="o">.</span><span class="n">rlen</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">read_stock</span>
</div>
<div class="viewcode-block" id="aligned_assembly.make_tree"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly.make_tree">[docs]</a>    <span class="k">def</span> <span class="nf">make_tree</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Function to construct an interval tree from reads aligning to a contig and return the interval tree.</span>

<span class="sd">        The interval tree stores nodes with properties start (start postition of interval), end (end position of interval) and other,</span>
<span class="sd">        which is a tuple of the mate pair name (string) and the strand alignment of the two paired reads (boolean).</span>

<span class="sd">        Arguments:</span>
<span class="sd">        ref: Reference ID of the contig for which the interval tree is to be constructed</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">bridges</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_stock</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span> 
        <span class="c"># insert first interval into tree</span>
        <span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">correct_strands</span> <span class="o">=</span> <span class="n">bridges</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">IntervalNode</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">e1</span><span class="p">,</span> <span class="n">other</span><span class="o">=</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">correct_strands</span><span class="p">))</span>

        <span class="c"># insert the rest of the intervals</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">correct_strands</span><span class="p">)</span> <span class="ow">in</span> <span class="n">bridges</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
            <span class="n">tree</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">other</span><span class="o">=</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">correct_strands</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">tree</span>
</div>
<div class="viewcode-block" id="aligned_assembly.get_read_mappings"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly.get_read_mappings">[docs]</a>    <span class="k">def</span> <span class="nf">get_read_mappings</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Function to calculate the fraction of reads pairs within a contig that align correctly to opposite strands.</span>

<span class="sd">        Return five arrays: the positions at which strand alignment was evaluated, the fraction correctly aligned, the fraction incorrectly aligned to the same strand, the unmapped </span>
<span class="sd">        fraction and the fraction that have some other alignment issue.</span>

<span class="sd">        Arguments:</span>
<span class="sd">        ref: the reference id of the contig to be evaulated</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">dump_val</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">step</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">same_strand</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">opp_strand</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">unmapped</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">other</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">good_ratio</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">unmapped_ratio</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">bad_ratio</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">other_ratio</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mini_pos</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">reads</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sam</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">reference</span> <span class="o">=</span> <span class="n">ref</span><span class="p">)</span>
        <span class="c"># not that iterating in this manner works because the bam file is sorted.</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reads</span><span class="p">):</span>    
            <span class="n">mini_pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">pos</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">mate_is_unmapped</span><span class="p">:</span>
                <span class="n">unmapped</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="p">((</span><span class="n">r</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">mate_is_reverse</span><span class="p">))</span> <span class="ow">or</span> <span class="p">((</span><span class="n">r</span><span class="o">.</span><span class="n">mate_is_reverse</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">)):</span>
                <span class="n">same_strand</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">elif</span><span class="p">((</span><span class="n">r</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">mate_is_reverse</span><span class="p">))</span> <span class="ow">or</span> <span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">mate_is_reverse</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">)):</span>
                <span class="n">opp_strand</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">other</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="n">dump_val</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">total</span> <span class="o">=</span> <span class="n">same_strand</span> <span class="o">+</span> <span class="n">opp_strand</span> <span class="o">+</span> <span class="n">unmapped</span> <span class="o">+</span> <span class="n">other</span>
                <span class="n">good_ratio</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">same_strand</span><span class="p">)</span> <span class="o">/</span> <span class="n">total</span><span class="p">)</span>
                <span class="n">bad_ratio</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">opp_strand</span><span class="p">)</span> <span class="o">/</span> <span class="n">total</span><span class="p">)</span>
                <span class="n">unmapped_ratio</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">unmapped</span><span class="p">)</span> <span class="o">/</span> <span class="n">total</span><span class="p">)</span>
                <span class="n">other_ratio</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">other</span><span class="p">)</span> <span class="o">/</span> <span class="n">total</span><span class="p">)</span>

                <span class="n">same_strand</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">opp_strand</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">unmapped</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">other</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">mini_pos</span><span class="p">))</span>
                <span class="n">mini_pos</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">positions</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">good_ratio</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">bad_ratio</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">unmapped_ratio</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">other_ratio</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="aligned_assembly.get_mapping_anomalies"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly.get_mapping_anomalies">[docs]</a>    <span class="k">def</span> <span class="nf">get_mapping_anomalies</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Function to determine the frequency of strand mapping anomalies across the entire genome assembly.</span>

<span class="sd">        Calls get_read_mappings for each contig larger than the aligned_assembly.min_size and returns:</span>
<span class="sd">        1) a dictionary with keys = contig reference IDs; values = list of positions and strand alignment ratios as described in get_read_mappings   </span>
<span class="sd">        2) a dictionary of anomalies wiht keys = contig reference IDs, values = [list of positions for which the ratio of correctly aligned strands &lt; 0.75 (currently hard-coded), corresponding ratio of correctly aligned strands]</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">mapping_ratios</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">anomalies</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">w</span><span class="p">,</span> <span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">refdict</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()):</span>
            <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_size</span><span class="p">:</span>
                <span class="c">#if length &lt; 1000000:</span>
                <span class="n">positions</span><span class="p">,</span> <span class="n">good_ratio</span><span class="p">,</span> <span class="n">bad_ratio</span><span class="p">,</span> <span class="n">unmapped_ratio</span><span class="p">,</span> <span class="n">other_ratio</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_read_mappings</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span>
                <span class="n">map_criterion</span> <span class="o">=</span> <span class="n">good_ratio</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">fraction</span>
                <span class="n">pos_anomalies</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">map_criterion</span><span class="p">]</span>
                <span class="n">map_anomalies</span> <span class="o">=</span> <span class="n">good_ratio</span><span class="p">[</span><span class="n">map_criterion</span><span class="p">]</span>
                <span class="n">mapping_ratios</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">positions</span><span class="p">,</span> <span class="n">good_ratio</span><span class="p">,</span> <span class="n">bad_ratio</span><span class="p">,</span> <span class="n">unmapped_ratio</span><span class="p">,</span> <span class="n">other_ratio</span><span class="p">]</span>
                <span class="n">anomalies</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pos_anomalies</span><span class="p">,</span> <span class="n">map_anomalies</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">mapping_ratios</span><span class="p">,</span> <span class="n">anomalies</span>
</div>
<div class="viewcode-block" id="aligned_assembly.get_size_anomalies"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly.get_size_anomalies">[docs]</a>    <span class="k">def</span> <span class="nf">get_size_anomalies</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Function to determine the frequency of insert size anomalies across the entire genome assembly.</span>

<span class="sd">        Calls probability_of_readlength for each contig larger than the aligned_assembly.min_size and returns:</span>
<span class="sd">        1) a dictionary with keys = contig reference IDs; values = array of Zscores as described in probability_of_readlength   </span>
<span class="sd">        2) a dictionary of anomalies wiht keys = contig reference IDs, values = [list of positions for which  abs(z-score) &gt; 2 (currently hard-coded), corresponding z-score value]</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">anomalies</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">zscores</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">all_probabilities</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">stock_probabilities</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">w</span><span class="p">,</span> <span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">refdict</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()):</span>
            <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_size</span><span class="p">:</span>
                <span class="c">#if length &lt; 1000000:</span>
                <span class="n">tree</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_tree</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span>
                <span class="n">positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">step</span><span class="p">,</span> <span class="n">length</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">window</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">step</span><span class="p">)</span>
                <span class="n">probabilities</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">pos</span> <span class="o">%</span> <span class="mi">10000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="k">print</span> <span class="n">pos</span>
                    <span class="n">bridges</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">find_intersections</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">pos</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">window</span><span class="p">,</span> <span class="n">pos</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">window</span><span class="p">))</span>
                    <span class="n">bridge_lengths</span><span class="p">,</span> <span class="n">strand_alignment</span> <span class="o">=</span> <span class="n">get_insertlengths</span><span class="p">(</span><span class="n">bridges</span><span class="p">)</span>
                    <span class="n">prob_lengths</span> <span class="o">=</span> <span class="n">probability_of_readlength</span><span class="p">(</span><span class="n">bridge_lengths</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">isize_mean</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">isize_MAD</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">prior</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span>
                    <span class="n">condition</span> <span class="o">=</span> <span class="n">strand_alignment</span> <span class="o">==</span> <span class="mi">1</span>
                    <span class="c">#print condition</span>
                    <span class="n">D</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">prob_lengths</span><span class="p">[</span><span class="n">condition</span><span class="p">])</span>
                    <span class="c">#D = np.sum(prob_lengths)</span>
                    <span class="c">#print &quot;D = &quot;, D, len(prob_lengths)#, prob_lengths</span>
                    <span class="n">probabilities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
                    <span class="n">all_probabilities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
                <span class="n">stock_probabilities</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">positions</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">probabilities</span><span class="p">)]</span>
        <span class="n">p_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">all_probabilities</span><span class="p">))</span>
        <span class="n">p_std</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">all_probabilities</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">ref</span><span class="p">,</span> <span class="p">[</span><span class="n">positions</span><span class="p">,</span> <span class="n">probs</span><span class="p">]</span> <span class="ow">in</span> <span class="n">stock_probabilities</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="n">zscore</span> <span class="o">=</span> <span class="p">(</span><span class="n">probs</span> <span class="o">-</span> <span class="n">p_mean</span><span class="p">)</span> <span class="o">/</span> <span class="n">p_std</span>
            <span class="c">#z_criterion = (zscore &gt; 4) | (zscore &lt; -4)</span>
            <span class="n">z_criterion</span> <span class="o">=</span> <span class="p">(</span><span class="n">zscore</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">threshold</span><span class="p">)</span>
            <span class="n">z_anomalies</span> <span class="o">=</span> <span class="n">zscore</span><span class="p">[</span><span class="n">z_criterion</span><span class="p">]</span>
            <span class="k">print</span> <span class="n">ref</span><span class="p">,</span> <span class="n">z_anomalies</span>
            <span class="n">pos_anomalies</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">z_criterion</span><span class="p">]</span>
            <span class="n">zscores</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">positions</span><span class="p">,</span> <span class="n">zscore</span><span class="p">]</span>
            <span class="c">#zscores[ref] = [positions, probs]</span>
            <span class="n">anomalies</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pos_anomalies</span><span class="p">,</span> <span class="n">z_anomalies</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">zscores</span><span class="p">,</span> <span class="n">anomalies</span><span class="p">,</span> <span class="n">tree</span>
</div>
<div class="viewcode-block" id="aligned_assembly.get_anomalies"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly.get_anomalies">[docs]</a>    <span class="k">def</span> <span class="nf">get_anomalies</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span> <span class="n">trim</span><span class="p">,</span> <span class="n">img_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Function to determine the frequency of anomalous mate pair behaviour across the entire genome assembly and return a dictionary where:</span>
<span class="sd">        key = contig reference IDs, </span>
<span class="sd">        value = list of postions within that contig where an assembly error is identified and the contig should be broken.</span>

<span class="sd">        Calls get_size_anomalies and get_mapping_anomalies for each contig larger than the aligned_assembly.min_size; makes a .csv file listing for each contig the positions of identified misassemblies and their corresponding anomalous scores.</span>

<span class="sd">        Arguments:</span>
<span class="sd">        outfile: name of file (including filepath) to store the list of contig misassemblies.</span>

<span class="sd">        Keyword Arguments:</span>
<span class="sd">        img_name: name of file (including filepath, not including filetype) to store plots of alignment quality</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">print</span> <span class="s">&quot;Anomaly detection&quot;</span>
        <span class="n">zscores</span><span class="p">,</span> <span class="n">size_anomalies</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_size_anomalies</span><span class="p">()</span>
        <span class="n">map_ratios</span><span class="p">,</span> <span class="n">map_anomalies</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_mapping_anomalies</span><span class="p">()</span>
        <span class="c">#print size_anomalies.keys()</span>
        <span class="c">#print map_anomalies.keys(), map_anomalies.items()</span>

        <span class="n">break_points</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">w</span><span class="p">,</span> <span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="p">[</span><span class="n">positions</span><span class="p">,</span> <span class="n">probs</span><span class="p">])</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">zscores</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()):</span>
            <span class="k">print</span> <span class="s">&quot;Writing D scores to file (</span><span class="si">%s</span><span class="s">)&quot;</span> <span class="o">%</span> <span class="n">ref</span>
            <span class="k">for</span> <span class="n">pos</span><span class="p">,</span> <span class="n">prob</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">probs</span><span class="p">):</span>
                <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s"> </span><span class="si">%s</span><span class="s"> </span><span class="si">%s</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">prob</span><span class="p">))</span>
            <span class="n">z_pos</span><span class="p">,</span> <span class="n">z_anomalies</span> <span class="o">=</span> <span class="n">size_anomalies</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span> <span class="c"># seems good</span>
            <span class="k">print</span> <span class="s">&quot;still good?&quot;</span>
            <span class="k">print</span> <span class="s">&quot;z_anomalies:&quot;</span><span class="p">,</span> <span class="n">z_anomalies</span>
            <span class="k">print</span> <span class="n">ref</span><span class="p">,</span> <span class="n">z_pos</span>
            <span class="n">map_positions</span><span class="p">,</span> <span class="n">good_ratio</span><span class="p">,</span> <span class="n">bad_ratio</span><span class="p">,</span> <span class="n">unmapped_ratio</span><span class="p">,</span> <span class="n">other_ratio</span> <span class="o">=</span> <span class="n">map_ratios</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span> 
            <span class="n">pos_anomalies</span><span class="p">,</span> <span class="n">map_anom</span> <span class="o">=</span> <span class="n">map_anomalies</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span>

            <span class="c">#ax2 = ax1.twinx()</span>
            <span class="c">#ax2.set_xlim([0, max(map_positions)])</span>
            <span class="c">#ax2.set_ylabel(&quot;Opposite Strand&quot;)</span>
            <span class="c">#lns2 = ax2.plot(map_positions, good_ratio, c=&quot;g&quot;, label=&quot;Opposite Strands&quot;)</span>
            <span class="c">#ax2.scatter(pos_anomalies, map_anom, c=&quot;r&quot;, marker=&quot;o&quot;)</span>
            <span class="c">#ax2.set_ylim([0, 1])</span>

            <span class="c">#lns = lns1</span>
            <span class="c">#labs = [l.get_label() for l in lns]</span>
            <span class="c">#plt.legend(lns, labs, loc=&quot;lower right&quot;)</span>
            <span class="n">anomaly_positions</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">z_pos</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
            <span class="k">print</span> <span class="n">ref</span><span class="p">,</span> <span class="n">anomaly_positions</span>

            <span class="n">break_points</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            
            <span class="c">#outfile.write(&quot;%s\n&quot; % ref)</span>
            <span class="c">#outfile.write(&quot;Position,Zscore,Fraction Opposite\n&quot;)</span>
            <span class="c">#for p1, z1 in zip(z_pos, z_anomalies):</span>
            <span class="c">#    outfile.write(&quot;%s,%s,-\n&quot; %(p1, z1))</span>
            <span class="c">#for p2, m in zip(pos_anomalies, map_anom):</span>
            <span class="c">#    outfile.write(&quot;%s,-,%s\n&quot; %(p2, m))</span>
            <span class="c">#outfile.write(&quot;\n&quot;)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">anomaly_positions</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">current</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">anomaly_positions</span><span class="p">)):</span>
                    <span class="n">anom_pos</span> <span class="o">=</span> <span class="n">anomaly_positions</span><span class="p">[</span><span class="n">p</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">current</span> <span class="o">==</span> <span class="p">[]:</span>
                        <span class="n">current</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">anom_pos</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">anom_pos</span> <span class="o">-</span> <span class="n">current</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">trim</span><span class="p">:</span>
                        <span class="c"># if anomalies are well separated flush current values to break_point</span>
                        <span class="c">#break_points[ref].append(np.mean(current))</span>
                            <span class="n">current</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">anom_pos</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">break_points</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">current</span><span class="p">))</span>
                            <span class="n">current</span> <span class="o">=</span> <span class="p">[</span><span class="n">anom_pos</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">current</span> <span class="o">!=</span> <span class="p">[]:</span>
                    <span class="n">break_points</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">current</span><span class="p">))</span>
                <span class="k">print</span> <span class="n">ref</span><span class="p">,</span> <span class="n">break_points</span><span class="p">[</span><span class="n">ref</span><span class="p">]</span>

            <span class="n">fig</span><span class="p">,</span> <span class="n">ax1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mf">0.15</span><span class="p">)</span>
            <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">map_positions</span><span class="p">)])</span>
            <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&quot;Position&quot;</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="mi">24</span><span class="p">)</span>
            <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&quot;Assembly Support&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">24</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s">&#39;both&#39;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s">&#39;major&#39;</span><span class="p">,</span> <span class="n">labelsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
            <span class="n">lns1</span> <span class="o">=</span> <span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">probs</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s">&quot;k&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s">&quot;Support&quot;</span><span class="p">)</span>
            <span class="c">#plt.ticklabel_format(style=&#39;sci&#39;, axis=&#39;x&#39;, scilimits=(0,0))</span>
            <span class="c">#ax1.set_xticks([0, 400000, 800000, 1200000])</span>
            <span class="n">anomalies_to_plot</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">break_points</span><span class="p">[</span><span class="n">ref</span><span class="p">])</span>
            <span class="c">#anomalies_y = np.zeros(len(anomalies_to_plot))</span>
            <span class="n">anomalies_y</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mi">10</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">anomalies_to_plot</span><span class="p">)</span>

            <span class="n">ax1</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">anomalies_to_plot</span><span class="p">,</span> <span class="n">anomalies_y</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s">&quot;r&quot;</span><span class="p">,</span> <span class="n">marker</span> <span class="o">=</span> <span class="s">&quot;o&quot;</span><span class="p">)</span>
            <span class="k">print</span> <span class="s">&quot;anomalies&quot;</span><span class="p">,</span> <span class="n">anomalies_to_plot</span>

            <span class="k">if</span> <span class="n">img_name</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">img_name</span> <span class="o">+</span> <span class="s">&quot;_</span><span class="si">%s</span><span class="s">.pdf&quot;</span> <span class="o">%</span> <span class="n">ref</span>
                <span class="c">#name = img_name + &quot;_%s.png&quot; % ref </span>
                <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">cla</span><span class="p">()</span>
        
        <span class="k">return</span> <span class="n">break_points</span>
</div>
<div class="viewcode-block" id="aligned_assembly.breakContigs_single"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly.breakContigs_single">[docs]</a>    <span class="k">def</span> <span class="nf">breakContigs_single</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span> <span class="n">breakpoints</span><span class="p">,</span> <span class="n">trim</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Function to break a contigs at positions identified as assembly errors and write a new fasta file containing all contigs (both altered and unaltered).</span>

<span class="sd">        Makes a single point break at the identified misassembly position</span>

<span class="sd">        Arguments:</span>
<span class="sd">        outfile: name of the new fasta file (including filepath)</span>
<span class="sd">        breakpoints: dictionary of misassemblies. key = contig reference ID, value = list of misassembly positions within the contig</span>
<span class="sd">        trim: distance, in bases, to trim from each each edge of a breakpoint to remove misassembly (integer)</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">breakpoints</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="k">print</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span>
        <span class="n">w</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">isize_mean</span><span class="p">)</span>
        <span class="n">newcontigs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">contig</span><span class="p">,</span> <span class="n">length</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">refdict</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="n">dna</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fasta</span><span class="p">[</span><span class="n">contig</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">contig</span> <span class="ow">in</span> <span class="n">breakpoints</span><span class="p">:</span>
                <span class="n">splits</span> <span class="o">=</span> <span class="n">breakpoints</span><span class="p">[</span><span class="n">contig</span><span class="p">]</span>
                <span class="n">splits</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
                <span class="n">prev</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">splits</span><span class="p">:</span>
                    <span class="k">print</span> <span class="n">s</span>
                    <span class="k">print</span> <span class="s">&quot;breaking contig&quot;</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">s</span> <span class="o">&gt;</span> <span class="n">trim</span><span class="p">)</span> <span class="ow">and</span> <span class="p">((</span><span class="n">length</span> <span class="o">-</span> <span class="n">s</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">trim</span><span class="p">):</span> 
                        <span class="n">newcontigs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dna</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">prev</span><span class="p">):</span><span class="nb">int</span><span class="p">(</span><span class="n">s</span><span class="p">)])</span>
                        <span class="n">prev</span> <span class="o">=</span> <span class="n">s</span>
                <span class="n">newcontigs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dna</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">prev</span><span class="p">):])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">newcontigs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>

        <span class="n">newcontigs</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">:</span> <span class="nb">cmp</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">y</span><span class="p">)),</span><span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="k">print</span> <span class="s">&quot;Writing new fasta...&quot;</span>
        <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">dna</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">newcontigs</span><span class="p">):</span>
            <span class="n">name</span> <span class="o">=</span> <span class="s">&quot;&gt;CONTIG_</span><span class="si">%d</span><span class="s">_length_</span><span class="si">%d</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">count</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">dna</span><span class="p">))</span>
            <span class="k">print</span> <span class="n">name</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="aligned_assembly.breakContigs_double"><a class="viewcode-back" href="../API.html#nxrepair.aligned_assembly.breakContigs_double">[docs]</a>    <span class="k">def</span> <span class="nf">breakContigs_double</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">outfile</span><span class="p">,</span> <span class="n">breakpoints</span><span class="p">,</span> <span class="n">trim</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Function to break a contigs at positions identified as assembly errors and write a new fasta file containing all contigs (both altered and unaltered).</span>

<span class="sd">        Makes a two-point break at the identified misassembly position, splitting at 5 Kb upstream and downstream of the misassembly and (currently) excluding the misassembled region.</span>

<span class="sd">        Arguments:</span>
<span class="sd">        outfile: name of the new fasta file (including filepath)</span>
<span class="sd">        breakpoints: dictionary of misassemblies. key = contig reference ID, value = list of misassembly positions within the contig</span>
<span class="sd">        trim: distance, in bases, to trim from each each edge of a breakpoint to remove misassembly (integer)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">breakpoints</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="k">print</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span>
        <span class="n">newcontigs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">contig</span><span class="p">,</span> <span class="n">length</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">refdict</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="n">dna</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fasta</span><span class="p">[</span><span class="n">contig</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">contig</span> <span class="ow">in</span> <span class="n">breakpoints</span><span class="p">:</span>
                <span class="n">splits</span> <span class="o">=</span> <span class="n">breakpoints</span><span class="p">[</span><span class="n">contig</span><span class="p">]</span>
                <span class="n">splits</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
                <span class="n">prev</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">splits</span><span class="p">:</span>
                    <span class="k">print</span> <span class="n">s</span>
                    <span class="k">print</span> <span class="s">&quot;breaking contig&quot;</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">s</span> <span class="o">-</span> <span class="n">prev</span> <span class="o">&gt;</span> <span class="n">trim</span><span class="p">)</span> <span class="ow">and</span> <span class="p">((</span><span class="n">length</span> <span class="o">-</span> <span class="n">s</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">trim</span><span class="p">):</span>
                        <span class="k">print</span> <span class="s">&quot;breaking here:&quot;</span><span class="p">,</span> <span class="n">s</span>
                        <span class="n">newcontigs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dna</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">prev</span><span class="p">):</span><span class="nb">int</span><span class="p">(</span><span class="n">s</span><span class="o">-</span><span class="n">trim</span><span class="p">)])</span>
                        <span class="c"># newcontigs.append(dna[int(s-5000):int(s+5000)])</span>
                        <span class="n">prev</span> <span class="o">=</span> <span class="n">s</span> <span class="o">+</span> <span class="n">trim</span>
                    <span class="c">#elif (s - prev &gt;= 10000) and ((length - s) &gt; 10000):</span>
                    <span class="c">#    prev = s + 10000 </span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">print</span> <span class="s">&quot;Too small!&quot;</span>
                <span class="n">newcontigs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dna</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">prev</span><span class="p">):])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">newcontigs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>

        <span class="n">newcontigs</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">:</span> <span class="nb">cmp</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">y</span><span class="p">)),</span><span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="k">print</span> <span class="s">&quot;Writing new fasta...&quot;</span>
        <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">dna</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">newcontigs</span><span class="p">):</span>
            <span class="n">name</span> <span class="o">=</span> <span class="s">&quot;&gt;CONTIG_</span><span class="si">%d</span><span class="s">_length_</span><span class="si">%d</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">count</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">dna</span><span class="p">))</span>
            <span class="k">print</span> <span class="n">name</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>    
</div></div>
<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="kn">import</span> <span class="nn">argparse</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s">&#39;Routine to identify and correct large-scale misassemblies in de novo assemblies&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;bam&#39;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&#39;bam&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;bam&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;fasta&#39;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&#39;fasta&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;scaffold fasta&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;outfile&#39;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&#39;outfile&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Output file name&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;newfasta&#39;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&#39;newfasta&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Fasta file for new contigs, including filepath&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-min_size&#39;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s">&#39;min_size&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Minimum contig size to analyse&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-img_name&#39;</span><span class="p">,</span> <span class="n">metavar</span> <span class="o">=</span><span class="s">&#39;img_name&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Name under which to save (optional) graphs of alignment quality. Default value: None (no graphs produced)&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-trim&#39;</span><span class="p">,</span> <span class="n">metavar</span> <span class="o">=</span><span class="s">&#39;trim&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Number of bases to trim from each side of an identified misassembly. Default value: 5000&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-T&#39;</span><span class="p">,</span> <span class="n">metavar</span> <span class="o">=</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span> <span class="o">-</span><span class="mf">4.0</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Threshold in Z score below which a misassembly is called. Default value: -4.0&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-step_size&#39;</span><span class="p">,</span> <span class="n">metavar</span> <span class="o">=</span><span class="s">&#39;step_size&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Step-size in bases to traverse contigs. Default value: 1000&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-window&#39;</span><span class="p">,</span> <span class="n">metavar</span> <span class="o">=</span><span class="s">&#39;window&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Window size across which bridging mate pairs are evaluated. Default value: 200&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-minmapq&#39;</span><span class="p">,</span> <span class="n">metavar</span> <span class="o">=</span><span class="s">&#39;minmapq&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Minimum MapQ value, above which a read pair is included in calculating population statistics. Default value: 40&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-maxinsert&#39;</span><span class="p">,</span> <span class="n">metavar</span> <span class="o">=</span><span class="s">&#39;maxinsert&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">30000</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Maximum insert size, below which a read pair is included in calculating population statistics. Default value: 30000&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-fraction&#39;</span><span class="p">,</span> <span class="n">metavar</span> <span class="o">=</span><span class="s">&#39;fraction&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Minimum fraction of read pairs with correct orientation to call support for the assembly. Default value: 0.75&#39;</span><span class="p">)</span>  
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&#39;-prior&#39;</span><span class="p">,</span> <span class="n">metavar</span> <span class="o">=</span><span class="s">&#39;prior&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&#39;Prior probablility that the insert size is anomalous. Default value: 0.01&#39;</span><span class="p">)</span>  


    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="n">f</span> <span class="o">=</span> <span class="n">aligned_assembly</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bam</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">fasta</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">min_size</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">step_size</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">window</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">minmapq</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">maxinsert</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">fraction</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">prior</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&quot;This is ok&quot;</span>
    
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">outfile</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">of</span><span class="p">:</span>
        <span class="n">bps</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">get_anomalies</span><span class="p">(</span><span class="n">of</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">trim</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">img_name</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">newfasta</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfasta</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">breakContigs_double</span><span class="p">(</span><span class="n">outfasta</span><span class="p">,</span> <span class="n">bps</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">trim</span><span class="p">)</span>

<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">misassembly_detection 1.0 documentation</a> &raquo;</li>
          <li><a href="index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Rebecca R. Murphy.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>