<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Tutorial: Misassembly Detection using Insert Probability &mdash; misassembly_detection 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="misassembly_detection 1.0 documentation" href="index.html" />
    <link rel="prev" title="Missassembly Detection With Python: API" href="API.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="API.html" title="Missassembly Detection With Python: API"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">misassembly_detection 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="tutorial-misassembly-detection-using-insert-probability">
<h1>Tutorial: Misassembly Detection using Insert Probability<a class="headerlink" href="#tutorial-misassembly-detection-using-insert-probability" title="Permalink to this headline">¶</a></h1>
<p>NxRepair is a python module that automatically detects large structural errors in de novo assemblies using Nextera mate pair reads. The decector will break a contig at the site of an identified misassembly and will generate a new fasta file containing both the corrected contigs and the correct, unaffected contigs.</p>
<div class="section" id="installing-nxrepair">
<h2>Installing NxRepair<a class="headerlink" href="#installing-nxrepair" title="Permalink to this headline">¶</a></h2>
<p>NxRepair program can be cloned from github:</p>
<div class="highlight-bash"><div class="highlight"><pre>git clone https://github.com/rebeccaroisin/nxrepair
</pre></div>
</div>
<p>NxRepair uses several other python libraries, which you will need to install. Specifically, you will need:</p>
<ul class="simple">
<li>scipy</li>
<li>numpy</li>
<li>matplotlib</li>
<li>pysam</li>
</ul>
<p>Scipy, numpy and matplotlib can be installed using your favourite package manager or from the Python Package Index (PyPI) using:</p>
<div class="highlight-bash"><div class="highlight"><pre>pip install numpy
pip install scipy
pip install matplotlib
</pre></div>
</div>
<p>If you don&#8217;t currently have any of these libraries, we recommend installing <a class="reference external" href="https://store.continuum.io/cshop/anaconda/">Anaconda</a>, a python distribution that includes all of these libraries, along with all major scientific and analytical python packages.</p>
<p>Pysam can also be installed using &#8220;pip install&#8221;, but errors are more common.</p>
<p>Assistance with installation errors for <a class="reference external" href="https://groups.google.com/forum/#!forum/pysam-user-group">pysam</a> can be found online. Note that the current version of pysam wraps samtools-0.1.19 and tabix-0.2.6.</p>
</div>
<div class="section" id="running-nxrepair">
<h2>Running NxRepair<a class="headerlink" href="#running-nxrepair" title="Permalink to this headline">¶</a></h2>
<p>NxRepair can be run from the command line.</p>
<div class="highlight-bash"><div class="highlight"><pre>python nxrepair.py aligned_matepairs.bam assemblyfasta.fasta error_locations.csv new_fasta.fasta
</pre></div>
</div>
<p>The required arguments are as follows:</p>
<ul class="simple">
<li>aligned_matepairs.bam: an indexed bam file of mate pair reads aligned to your assembly;</li>
<li>assemblyfasta.fasta: the fasta file containing your contigs;</li>
<li>outfile: the name of a csv file in which to store the Z scores generated by NxRepair;</li>
<li>newfasta: filename of the fasta file that will hold the new contigs following analysis.</li>
</ul>
<p>There are also several optional arguments, which can be used to tune the NxRepair analysis. These are described in the table below.</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="13%" />
<col width="79%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Default Value</th>
<th class="head">Meaning</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>imgname</td>
<td>None</td>
<td>Prefix under which to save plots.</td>
</tr>
<tr class="row-odd"><td>maxinsert</td>
<td>30000</td>
<td>Maximum insert size, below which a read pair is included in calculating population statistics.</td>
</tr>
<tr class="row-even"><td>minmapq</td>
<td>40</td>
<td>Minimum MapQ value, above which a read pair is included in calculating population statistics.</td>
</tr>
<tr class="row-odd"><td>minsize</td>
<td>10000</td>
<td>Minimum contig size to analyse.</td>
</tr>
<tr class="row-even"><td>prior</td>
<td>0.01</td>
<td>Prior probablility that the insert size is anomalous.</td>
</tr>
<tr class="row-odd"><td>stepsize</td>
<td>1000</td>
<td>Step-size in bases to traverse contigs.</td>
</tr>
<tr class="row-even"><td>trim</td>
<td>5000</td>
<td>Number of bases to trim from each side of an identified misassembly.</td>
</tr>
<tr class="row-odd"><td>T</td>
<td>-4.0</td>
<td>Threshold in Z score (standard deviations from the mean) below which a misassembly is called.</td>
</tr>
<tr class="row-even"><td>window</td>
<td>200</td>
<td>Window size across which bridging mate pairs are evaluated.</td>
</tr>
</tbody>
</table>
<p>Optional arguments are called from the commmand line, as shown in the example below:</p>
<div class="highlight-bash"><div class="highlight"><pre>python nxrepair.py aligned_matepairs.bam assemblyfasta.fasta error_locations.csv new_fasta.fasta -minsize 20000 -trim 4000 -T -5.0
</pre></div>
</div>
<p>The program will parse a bam file of reads aligned to your de novo assembly. Each contig that is larger than the min_size parameter will be analysed for potential structural misassemblies. When the program completes, a minimum of two new files will be generated:</p>
<ol class="arabic simple">
<li>A new fastafile, specified by newfasta, that contains the improved contigs of the de novo assembly.</li>
<li>A csv file that identifies the exact position where altered contigs were broken.</li>
<li>If the optional argument -img_name was included, for each contig analysed, a plot will be generated showing the insert size distribution and directionality across the contig, with anomalous regions highlighted. These plots will be saved in the folder specified by img_name</li>
</ol>
<p>Outputs 2 and 3 can allow identification of further, smaller structural misassemblies, as well as enabling verification of detected misassemblies using IGV.</p>
</div>
<div class="section" id="how-does-it-work">
<h2>How Does it Work?<a class="headerlink" href="#how-does-it-work" title="Permalink to this headline">¶</a></h2>
<p>NxRepair evaluates the insert sizes of mate pairs aligned across a contig. Regions of the contig that have unusual insert sizes, where few reads are aligned, or where a large fraction of the mate pairs have incorrect orientation are flagged as potentially anomalous based on a simple probabilistic model of the mate-pair size distribution. Where there is strong evidence that a region is misassembled, the contig will be broken into two pieces and 5 Kb of erroneous assemby will be trimmed from both sides of the break.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Tutorial: Misassembly Detection using Insert Probability</a><ul>
<li><a class="reference internal" href="#installing-nxrepair">Installing NxRepair</a></li>
<li><a class="reference internal" href="#running-nxrepair">Running NxRepair</a></li>
<li><a class="reference internal" href="#how-does-it-work">How Does it Work?</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="API.html"
                        title="previous chapter">Missassembly Detection With Python: API</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/tutorial.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="API.html" title="Missassembly Detection With Python: API"
             >previous</a> |</li>
        <li><a href="index.html">misassembly_detection 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Rebecca R. Murphy.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>